File: app/src/main/java/igrek/songbook/service/chords/LyricsManager.java
package igrek.songbook.service.chords;

import android.graphics.Paint;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.domain.crd.CRDModel;
import igrek.songbook.domain.crd.CRDParser;
import igrek.songbook.service.autoscroll.AutoscrollService;
import igrek.songbook.service.chords.transpose.ChordsTransposerManager;
import igrek.songbook.service.preferences.PreferencesDefinition;
import igrek.songbook.service.preferences.PreferencesService;

public class LyricsManager {
	
	@Inject
	Lazy<ChordsTransposerManager> chordsTransposerManager;
	@Inject
	Lazy<AutoscrollService> autoscrollService;
	@Inject
	PreferencesService preferencesService;
	
	private CRDParser crdParser;
	private CRDModel crdModel;
	private int screenW = 0;
	private Paint paint;
	private float fontsize;
	private String originalFileContent;
	
	public LyricsManager() {
		DaggerIoc.getFactoryComponent().inject(this);
		crdParser = new CRDParser();
		loadPreferences();
	}
	
	private void loadPreferences() {
		fontsize = preferencesService.getValue(PreferencesDefinition.fontsize, Float.class);
	}
	
	private void reset() {
		chordsTransposerManager.get().reset();
		autoscrollService.get().reset();
	}
	
	public void load(String fileContent, Integer screenW, Integer screenH, Paint paint) {
		reset();
		originalFileContent = fileContent;
		
		if (screenW != null) {
			this.screenW = screenW;
		}
		if (paint != null) {
			this.paint = paint;
		}
		
		crdParser = new CRDParser();
		
		parseAndTranspose(originalFileContent);
	}
	
	
	public void reparse() {
		parseAndTranspose(originalFileContent);
	}
	
	public CRDModel getCRDModel() {
		return crdModel;
	}
	
	public float getFontsize() {
		return fontsize;
	}
	
	public void setFontsize(float fontsize) {
		if (fontsize < 1)
			fontsize = 1;
		this.fontsize = fontsize;
	}
	
	private void parseAndTranspose(String originalFileContent) {
		String transposedContent = chordsTransposerManager.get()
				.transposeContent(originalFileContent);
		crdModel = crdParser.parseFileContent(transposedContent, screenW, fontsize, paint);
	}
	
}

File: app/src/main/java/igrek/songbook/songpreview/lyrics/LyricsManager.kt
package igrek.songbook.songpreview.lyrics

import android.graphics.Paint
import dagger.Lazy
import igrek.songbook.chords.transpose.ChordsTransposerManager
import igrek.songbook.dagger.DaggerIoc
import igrek.songbook.settings.chordsnotation.ChordsNotation
import igrek.songbook.settings.preferences.PreferencesService
import igrek.songbook.settings.preferences.PreferencesState
import igrek.songbook.settings.theme.LyricsThemeService
import igrek.songbook.songpreview.autoscroll.AutoscrollService
import igrek.songbook.system.WindowManagerService
import javax.inject.Inject

class LyricsManager {

    @Inject
    lateinit var chordsTransposerManager: Lazy<ChordsTransposerManager>
    @Inject
    lateinit var autoscrollService: Lazy<AutoscrollService>
    @Inject
    lateinit var lyricsThemeService: LyricsThemeService
    @Inject
    lateinit var windowManagerService: WindowManagerService
    @Inject
    lateinit var preferencesService: PreferencesService
    @Inject
    lateinit var preferencesState: PreferencesState

    private var lyricsParser: LyricsParser? = null
    var crdModel: LyricsModel? = null
        private set
    private var screenW = 0
    private var paint: Paint? = null
    private var originalFileContent: String? = null

    private var restoreTransposition
        get() = preferencesState.restoreTransposition
        set(value) {
            preferencesState.restoreTransposition = value
        }

    init {
        DaggerIoc.factoryComponent.inject(this)
    }

    private fun normalizeContent(content: String): String {
        return content
                .replace("\r", "")
                .replace("\t", " ")
                .replace("\u00A0", " ") // NO-BREAK SPACE (0xC2 0xA0)
    }

    fun load(fileContent: String, screenW: Int?, paint: Paint?, initialTransposed: Int, srcNotation: ChordsNotation) {
        chordsTransposerManager.get().run {
            val transposed = when {
                restoreTransposition -> initialTransposed
                else -> 0
            }
            reset(transposed, srcNotation)
        }
        autoscrollService.get().reset()

        originalFileContent = normalizeContent(fileContent)

        if (screenW != null)
            this.screenW = screenW
        if (paint != null)
            this.paint = paint

        val typeface = lyricsThemeService.fontTypeface.typeface
        val chordsEndOfLine = lyricsThemeService.chordsEndOfLine
        val chordsAbove = lyricsThemeService.chordsAbove
        lyricsParser = LyricsParser(typeface, chordsEndOfLine, chordsAbove)

        parseAndTranspose(originalFileContent!!)
    }

    fun onPreviewSizeChange(screenW: Int, paint: Paint?) {
        this.screenW = screenW
        this.paint = paint
        reparse()
    }

    fun reparse() {
        parseAndTranspose(originalFileContent!!)
    }

    private fun parseAndTranspose(originalFileContent: String) {
        val transposedContent = chordsTransposerManager.get()
                .transposeContent(originalFileContent)
        val realFontsize = windowManagerService.dp2px(lyricsThemeService.fontsize)
        crdModel = lyricsParser?.parseFileContent(transposedContent, screenW.toFloat(), realFontsize, paint!!)
    }

}
